base/show.jl

```diff
@@ -266,7 +274,14 @@ function show_type_name(io::IO, tn::TypeName)
     elseif globfunc
         print(io, "typeof(")
     end
-    if isdefined(tn, :module) && !(is_exported_from_stdlib(sym, tn.module) || (tn.module === Main && !hidden))
+    # Print module prefix unless type is:
+    # - exported from standard library
+    # - reachable from module passed to IOContext
+    # - defined in Main and not hidden
+    if isdefined(tn, :module) &&
+        !(is_exported_from_stdlib(sym, tn.module) ||
+          is_reachable(sym, tn.module, get(io, :module, nothing)) ||
+          (tn.module === Main && !hidden))
```

Comment:
I feel like it might be better to *only* consider visibility. Doesn't that cover the other cases in the typical situations where we are displaying from `Main` and the `Base` has been imported? If the user asks to display something with respect to another module, wouldn't we want to only shorten names that are actually visible in that module? Why are `Main` or stdlib relevant there?

Reply:
I've just tried it, and it triggers a lot of test failures all over the place. That made me realize the `IOContext` approach implies hardcoding `:module => Main` in many places. For example, `stringmime` calls `_textreprmime(m::MIME, x) = sprint(verbose_show, m, x)`. To omit the module prefix, we need to create the `IOBuffer` by hand to pass it an adequate `IOContext`. There are many examples like that.

Do we really want to add this kind of thing all over the place? I'm asking before spending time to find them all. The problem is that in the end many functions do not take a `IO` object, so they will hardcode `Main`, which kind of defeats the use of `IOContext`. Is that still worth it, rather than hardcoding `Main` in a single central place?

Reply:
Actually, a good intermediate solution would be to default to `Main` (i.e. use `get(io, :module, Main)` instead of `get(io, :module, nothing)`), but still allow passing a different `:module`. That way by default we get reasonably short type names without hardcoding `Main` in many places, but it can still be customized.

Reply:
That sounds like a good approach: `Main` as the default `nothing` as a value that can be passed to have no context module. Another thing that occurs to me is allowing an array of modules to look in and having `[]` indicate not to look anywhere, defaulting to `[Main]`. But it would also make sense to allow a single module as a shortcut in that case. Maybe too complicated as a starting point.